lib/astarte/client/appengine.ex

```diff
@@ -46,20 +46,24 @@ defmodule Astarte.Client.AppEngine do
     end
   end
 
-  defp fetch_jwt(opts) do
-    with :error <- Keyword.fetch(opts, :jwt) do
-      opts
-      |> Keyword.get(:private_key)
-      |> generate_jwt()
+  defp fetch_or_generate_jwt(opts) when is_list(opts) do
+    with :error <- Keyword.fetch(opts, :jwt),
+         :error <- generate_jwt(opts) do
+      {:error, :missing_jwt_and_private_key}
```

Comment:
You're assuming `generate_jwt` can return only `:error`, but I think it can also return some other kind of error (e.g. `Credentials.to_jwt/2` calls `Joken.generate_and_sign`, which can return `{:error, reason}`).
So I think you should just return what `generate_jwt` returns here, and return `{:error, :missing_jwt_and_private_key}` in `generate_jwt` instead:
```elixir
  defp fetch_or_generate_jwt(opts) when is_list(opts) do
    with :error <- Keyword.fetch(opts, :jwt) do
      generate_jwt(opts)
    end
  end

  defp generate_jwt(opts) when is_list(opts) do
    case Keyword.fetch(opts, :private_key) do
      {:ok, private_key} ->
        opts
        |> Keyword.get(:jwt_opts, [])
        |> Keyword.put_new(:expiry, @jwt_expiry)
        |> Credentials.appengine_all_access_credentials()
        |> Credentials.to_jwt(private_key)

      _ ->
        {:error, :missing_jwt_and_private_key}
    end
  end
```